lib/builders.nix

```diff
@@ -1,135 +1,160 @@
+                    fi
+                    return
                 fi
-                return
-            fi
-
-            # Target is a directory
-            while IFS= read -r -d "" file; do
-                tmp=$(mktemp -u "$file.XXXX")
-                # Ignore invalid lua files
-                if ${luajit} -bd -- "$file" "$tmp"; then
-                    mv "$tmp" "$file"
-                else
-                    echo "WARNING: Ignoring byte compiling error for '$file' lua file" >&2
-                fi
-            done < <(find "$out" -type f,l -name "*.lua" -print0)
-        }
-
-        postFixupHooks+=(byteCompileLuaPostFixup)
-      ''
-  );
-
-  /*
-    Returns an overridden derivation with all lua files byte compiled.
-
-    # Type
 
-    ```
-    byteCompileLuaDrv :: Derivation -> Derivation
-    ```
-
-    # Arguments
-
-    - [drv] Input derivation
-  */
-  byteCompileLuaDrv =
-    drv:
-    drv.overrideAttrs (
-      prev:
-      {
-        nativeBuildInputs = prev.nativeBuildInputs or [ ] ++ [ byteCompileLuaHook ];
-      }
-      // lib.optionalAttrs (prev ? buildCommand) {
-        buildCommand = ''
-          ${prev.buildCommand}
-          runHook postFixup
-        '';
+                # Target is a directory
+                while IFS= read -r -d "" file; do
+                    tmp=$(mktemp -u "$file.XXXX")
+                    # Ignore invalid lua files
+                    if ${luajit} -bd -- "$file" "$tmp"; then
+                        mv "$tmp" "$file"
+                    else
+                        echo "WARNING: Ignoring byte compiling error for '$file' lua file" >&2
+                    fi
+                done < <(find "$out" -type f,l -name "*.lua" -print0)
+            }
+
+            postFixupHooks+=(byteCompileLuaPostFixup)
+          ''
+      );
+
+    /*
+      Returns an overridden derivation with all lua files byte compiled.
+
+      # Type
+
+      ```
+      byteCompileLuaDrvWith :: pkgs -> Derivation -> Derivation
+      ```
+
+      # Arguments
+
+      - [pkgs] A nixpkgs instance
+      - [drv] Input derivation
+    */
+    byteCompileLuaDrvWith =
+      pkgs: drv:
+      drv.overrideAttrs (
+        prev:
+        {
+          nativeBuildInputs = prev.nativeBuildInputs or [ ] ++ [
+            (lib.nixvim.builders.byteCompileLuaHookWith pkgs)
+          ];
+        }
+        // lib.optionalAttrs (prev ? buildCommand) {
+          buildCommand = ''
+            ${prev.buildCommand}
+            runHook postFixup
+          '';
+        }
+      );
+  };
+
+  # Curry a nixpkgs instance into the above functions, dropping the `With` suffix
+  withPkgs =
+    pkgs:
+    lib.concatMapAttrs (
+      name: fn:
+      let
+        match' = builtins.match "(.*)With" name;
+        name' = builtins.head match';
+      in
+      lib.optionalAttrs (match' != null) {
+        ${name'} = fn pkgs;
       }
-    );
-}
+    ) builders;
+in
+builders // { inherit withPkgs; }
```

Comment:
could this be simplified into just having the withPkgs in the let block ?

Reply:
We could do it recursively, true. The `builtins.match` should already filter out `withPkgs`

Reply:
I've updated it to be recursive. The diff is smaller now because the indentation is no longer changed.

Are you still happy?

Reply:
Looks better, at least to me.